### PR TITLE
Added development best practices

### PR DESCRIPTION
**Pull request by pramodin**
@sashadt please review and merge

This is from the discussion in the mailing list below. The document was available for review by everyone. There were comments from different folks and they were addressed.

https://www.mail-archive.com/dev%40apex.apache.org/msg01027.html

### CODE REVIEW

**Comment by pramodin** [comment]
@sashadt @tweise this is from the shared doc that folks looked, commented on and feedback incorporated.

**Comment by tweise** [comment]
There is only a link change, did you forget to add the new doc?

**Comment by tweise** [comment]
The operator and application developer guides are part of the core documentation set, I think we should add this document there as well. @amberarrow please also review

File: docs/development_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+* Sometimes operators need to perform some tasks just before checkpointing. For example, filesystem operators may want to flush the files just before checkpoint so they can be sure that all pending data is written to disk and no data is lost if there is an operator failure just after the checkpoint and the operator restarts from the checkpoint. To do this the operator would implement the same *CheckpointNotificationListener* interface and implement the **beforeCheckpoint** method where it can do these tasks.
+* If the operator state is going to be large, checkpointing the entire state each time becomes unviable. Furthermore, the amount of state to keep could be larger than the amount of physical memory available. In those cases operator developers should use more efficient ways of maintaining state such as using **ManagedState** utility provided by the platform which uses a combination of in memory and disk cache to efficiently store and retrieve data in a performant and fault tolerant way. There are operators in the platform that use *ManagedState* and can be used as a reference on how to use this utility such as Dedup or Join operators.
```
Comment (tweise) [inline docs/development_best_practices.md]: I would refer to this overall as incremental checkpointing.

**Comment by tweise** [comment]
Should there be a section about testing?

**Comment by pramodin** [comment]
@tweise makes sense will move it to core

File: docs/development_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Development Best Practices
+
+This document describes the best practices to follow when developing operators or other application components such as partitoners, stream codecs etc.
+
+## Operators
+
+These are general guidelines for all operators and the next sections talk about special considerations for input and output operators.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: "... operators and the next ..." reads better as "... operators; the following ...

File: docs/development_best_practices.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	* An optional concrete operator also in the library that extends the abstract operator and provides commonly used schema types such as strings, byte[] or POJOs.
+* Follow the conventions of the life cycle methods
+	* Do one time initialization of entities that apply for entire lifetime of the operator in the **setup** method, e.g., factory initializations. Initializations in **setup** are done in the container where the operator is deployed. Initializations in the constructor are discouraged as they lead to extra garbage in memory. This reason is as follows. The operator is instantiated on the client from where the application is being launched, then serialized and made available for launch. So the constructor is first called on the client and if it were to initialize any of the fields, that state would be saved during serialization. In the launch process the operator is deserialized. This would invoke the constructor again, which will initialize the fields but their state will get overwritten by the serialized state and the initial values would become garbage in memory.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: for entire => for the entire
This reason is as follows. => The reason is as follows:
is being launched => is launched
Initializations in the constructor are discouraged => Initializations of non-transient fields are discouraged
{seems like this also applies to initialization at the definition point ?}

File: docs/development_best_practices.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.  
+	* Perform periodic tasks based on processing time in application windows
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Not sure what this means -- some additional explanation and/or an example of such a "periodic task" would be helpful.

File: docs/development_best_practices.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	* Perform initializations needed for each application window in **beginWindow**
+	* Perform aggregations needed for each application window  in **endWindow**
```
Comment (amberarrow) [inline docs/development_best_practices.md:17]: Seems like this only applies to some aggregations (like average) and not to others (such as sum). Clarifying that would avoid confusion.

File: docs/development_best_practices.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	* Teardown of live entities (inverse of tasks performed during activate) should be in the **deactivate** method
+	* Teardown of lifetime entities (those initialized in setup method) should happen in the **teardown** method
+	* If the operator implementation is not finalized mark it with the **@Evolving** annotation
+* If the operator needs to perform operations based on event time of the individual tuples and not the processing time, extend and use the **WindowedOperator**. Refer to documentation of that operator for details on how to use it.
+* If an operator needs to do some work when it is not receiving any input, it should implement **IdleTimeHandler** interface. This interface contains **handleIdleTime** method which will be called whenever the platform isn’t doing anything else and the operator can do the work in this method. If for any reason the operator is not going to do any work in this method, it should sleep for a small amount of time such as that specified by the **SPIN_MILLIS** attribute otherwise it will cause a busy wait.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Is it worth adding a caution that work done within this method should not take excessively long, because otherwise the operator may be deemed blocked and killed ?

File: docs/development_best_practices.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+* Often operators have customizable parameters such as information about locations of external systems or parameters that modify the behavior of the operator. Users should be able to specify these easily without having to change source code. This can be done by making them properties of the operator because they can then be initialized from external properties files.
+	* Where possible default values should be provided for the properties in the source code.
+	* Validation rules should be specified for the properties using javax constraint validations that check whether the values specified for the properties are in the correct format, range or other operator requirements. Required properties should have at least a @NotNull validation specifying that they have to be specified by the user.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: "@NotNull"  => "**@NotNull**" {needs emboldening}

File: docs/development_best_practices.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+### Checkpointing
+
+Checkpointing is a process of snapshotting the state of an operator and saving it so that in case of failure the state can be used to restore the operator and continue processing. It is automatically performed by the platform at a configurable interval. All operators in the application are checkpointed in a distributed fashion, thus allowing the entire state of the application to be saved and available for recovery if needed. Here are some things to remember when it comes to checkpointing
```
Comment (amberarrow) [inline docs/development_best_practices.md]: restore the operator => restore the operator to the prior state
comes to checkpointing => comes to checkpointing:

File: docs/development_best_practices.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+* The process of checkpointing involves snapshotting the state by serializing the operator and saving it to a store. This is done using a **StorageAgent**. By default a *StorageAgent* is already provided by the platform and it is called **AsyncFSStorageAgent**. It serializes the operator using Kryo and saves the serialized state asynchronously to a filesystem such as HDFS. There are other implementations of *StorageAgent* available such as **GeodeKeyValueStorageAgent** that stores the serialized state in Geode which is an in-memory replicated data grid.
+* All variables in the operator not marked transient and final are saved so any variables in the operator that are not part of the state should be marked transient. Specifically any variables like connection objects, i/o streams, ports are transient, because they need to be setup again on failure recovery.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: not marked transient and final => marked neither transient nor final
{Adding an explicit caution that variables marked final are not saved may be worth doing since a common pattern is to define a collection as final but the content of the collection continues to accumulate state}

File: docs/development_best_practices.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+* If the operator does not keep any state between windows, mark it with the **@Stateless** annotation. This results in efficiencies during checkpointing and recovery. The operator will not be checkpointed and always restored to the initial state
```
Comment (amberarrow) [inline docs/development_best_practices.md]: and always => and is always

File: docs/development_best_practices.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+* The checkpoint interval can be set using the **CHECKPOINT\_WINDOW\_COUNT** attribute which specifies the interval in terms of number of streaming windows.
```
Comment (amberarrow) [inline docs/development_best_practices.md:34]: Not clear if underscores need to be escaped (the one in SPIN_MILLIS above is not)

File: docs/development_best_practices.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+* If the correct functioning of the operator requires the **endWindow** method be called before checkpointing can happen, then the checkpoint interval should align with application window interval i.e., it should be a multiple of application window interval. In this case the operator should be marked with **OperatorAnnotation** and **checkpointableWithAppWindow** set to false. If the window intervals are configured by the user and they don’t align, it will result in a DAG validation error and application won’t launch.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: checkpointableWithAppWindow => checkpointableWithinAppWindow

File: docs/development_best_practices.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+* If the operator state is going to be large, checkpointing the entire state each time becomes unviable. Furthermore, the amount of state to keep could be larger than the amount of physical memory available. In those cases operator developers should use more efficient ways of maintaining state such as using **ManagedState** utility provided by the platform which uses a combination of in memory and disk cache to efficiently store and retrieve data in a performant and fault tolerant way. There are operators in the platform that use *ManagedState* and can be used as a reference on how to use this utility such as Dedup or Join operators.
+
+## Input Operators
+
+Input operators have additional requirements
```
Comment (amberarrow) [inline docs/development_best_practices.md]: requirements => requirements:

File: docs/development_best_practices.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+* The **emitTuples** method is called by platform giving it an opportunity to emit some data. There are some guidelines on how to implement this method that are important
```
Comment (amberarrow) [inline docs/development_best_practices.md]: There are .... important => There are some important guidelines on how to implement this method:

File: docs/development_best_practices.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+	* This should not be a blocking method and should return in a reasonable time that is less than the streaming window size (which is 500ms by default).
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Doesn't this proviso also apply to most operator callbacks like beginWindow(), endWindow(), handleIdleTime() etc. ?

File: docs/development_best_practices.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+	* If the operator needs to interact with external systems to obtain data and this can potentially take a long time, then this should be performed asynchronously in a different thread. Refer to the threading section below for the guidelines when using threading.
+	* The method can emit more than one data tuples.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: more than one data tuples. => zero, one or multiple data tuples.
{alternatively, "any number of data tuples."}

File: docs/development_best_practices.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+### Idempotence
+
+Many applications write data to external systems using output operators. To ensure that data is present exactly once in the external system even in a failure recovery scenario, the output operators expect the replayed windows during recovery contain the same data as before the failure. This is called idempotency. Since operators within the DAG are merely responding to input data provided to them by the upstream operators and the input operator has no upstream operator, the responsibility of idempotent replay falls on the input operators.
+
+* For idempotent replay of data, the operator needs to store some meta-information for every window that would allow it to identify what data was sent in that window. This is called the idempotent state.
+	* If the external source of the input operator allows replayability, this could be information such as offset of last piece of data in the window, an identifier of the last piece of data itself or number of data tuples sent.
+	* However if the external source does not allow replayability from an operator specified point, then the entire data sent within the window may need to be persisted by the operator
```
Comment (amberarrow) [inline docs/development_best_practices.md]: operator => operator.

File: docs/development_best_practices.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+* The platform provides a utility called *WindowDataManager* to allow operators to save and retrieve idempotent state every window. Operators should use this to implement idempotency.
+
+## Output Operators
+
+Output operators typically connect to external storage systems such as filesystems, databases or key value stores to store data.
+
+* In some situations, the external systems may not be functioning in a reliable fashion. They may be having prolonged outages or performance problems. If the operator is being designed to work in such environments, it needs to be able to to handle these problems gracefully and not block the DAG or fail. In these scenarios the operator should cache the data into a local store such as HDFS and interact with external systems in a separate thread so as to not have problems in the operator lifecycle thread. The Reconciler pattern and operators that implement the pattern are available in the library for reference.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: functioning in a reliable fashion => reliable
The Reconciler pattern and operators that implement the pattern ... => This mode of operation is sometimes referred to as the Reconciler pattern; operators that implement it ...

File: docs/development_best_practices.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+### End-to-End Exactly Once
+
+When output operators store data in external systems, it is important that they do not lose data or write duplicate data when there is a failure event and the DAG recovers from that failure. In failure recovery, the windows from the previous checkpoint are replayed and the operator receives this data again. The operator should ensure that it does not write this data again. Operator developers should figure out how to do this specifically for the operators they are developing depending on the logic of the operators. Below are examples of how a couple of existing output operators do this for reference.
+
+* File output operator that writes data to files keeps track of the file lengths in the state. These lengths are checkpointed and restored on failure recovery. On restart, the operator truncates the file to the offset equal to the recovered length. This makes the data in the file same as it was at the time of checkpoint before the failure. The operator now writes the replayed data from the checkpoint in regular fashion as any other data. This ensures no data is lost or duplicated in the file.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: truncates the file to the offset equal to the recovered length => truncates the file to the recovered length

File: docs/development_best_practices.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+## Partitioning
+
+Partitioning allows an operation to be scaled to handle more pieces of data than before but with a similar SLA. This is done by creating multiple instances of an operator and distributing the data among them. Input operators can also be partitioned to stream more pieces of data into the application. Apex platform provides a lot of flexibility and options for partitioning. Partitioning can happen once at startup or can be dynamically changed anytime while the application is running, and it can be done in a stateless or stateful way by distributing state from the old partitions to new partitions.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: more pieces of data => more data {2 places}
similar SLA => similar performance characteristics by replicating operators
Apex platform => The Apex platform

File: docs/development_best_practices.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+In Apex, the responsibility for partitioning is shared among different entities. These are
```
Comment (amberarrow) [inline docs/development_best_practices.md]: These are => These are:

File: docs/development_best_practices.md [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+ 
+1. A **partitioner** that specifies *how* to partition the operator, specifically it takes an old set of partitions and creates a new set of partitions. At the start of the application the old set has one partition and the partitioner can return more than one partitions to start the application with multiple partitions. The partitioner can use any custom JAVA logic to determine the number of new partitions, set their initial state as a brand new state or by deriving it from the state of the old partitions, and specify how the data gets distributed between the partitions.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: custom JAVA logic => custom logic
between the partitions => among the new partitions.
Worth adding: The old and new set of partitions may intersect, i.e. some old partitions may be preserved unchanged in the new set if desired.

File: docs/development_best_practices.md [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+## Input Operators
+
+Input operators have additional requirements
+
+* The **emitTuples** method is called by platform giving it an opportunity to emit some data. There are some guidelines on how to implement this method that are important
+	* This should not be a blocking method and should return in a reasonable time that is less than the streaming window size (which is 500ms by default).
+	* If the operator needs to interact with external systems to obtain data and this can potentially take a long time, then this should be performed asynchronously in a different thread. Refer to the threading section below for the guidelines when using threading.
+	* The method can emit more than one data tuples.
+
+### Idempotence
```
Comment (amberarrow) [inline docs/development_best_practices.md:49]: Suggest changing the organization of this section to be more uniform: Currently, the emitTuples piece has no subtitle but Idempotence does.

File: docs/development_best_practices.md [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+	* However if the external source does not allow replayability from an operator specified point, then the entire data sent within the window may need to be persisted by the operator
+* The platform provides a utility called *WindowDataManager* to allow operators to save and retrieve idempotent state every window. Operators should use this to implement idempotency.
+
+## Output Operators
+
+Output operators typically connect to external storage systems such as filesystems, databases or key value stores to store data.
+
+* In some situations, the external systems may not be functioning in a reliable fashion. They may be having prolonged outages or performance problems. If the operator is being designed to work in such environments, it needs to be able to to handle these problems gracefully and not block the DAG or fail. In these scenarios the operator should cache the data into a local store such as HDFS and interact with external systems in a separate thread so as to not have problems in the operator lifecycle thread. The Reconciler pattern and operators that implement the pattern are available in the library for reference.
+
+### End-to-End Exactly Once
```
Comment (amberarrow) [inline docs/development_best_practices.md:64]: Here too, the top-level organization is somewhat asymmetric: the first part about the Reconciler has no subtitle but the second part does. Not sure how to reorg though ....

File: docs/development_best_practices.md [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+5. The Apex *engine* that brings everything together and effects the partitioning.
+
+Since partitioning is critical for scalability of applications, operators must support it. There should be a strong reason for an operator to not support partitioning, such as, the logic performed by the operator not lending itself to parallelism. In order to support partitioning, an operator developer, apart from developing the functionality of the operator, may also need to provide a partitioner, stats listener and supporting code in the operator as described in the steps above. The next sections delve into this. 
+
+### Out of the box partitioning
+
+Apex comes with some built-in partitioning utilities that can be used in certain scenarios.
+
+* **StatelessPartitioner** in Malhar provides a default partitioner, which can be used in certain conditions. If the operator satisfies these conditions, no special partitioning code is needed . These are
```
Comment (amberarrow) [inline docs/development_best_practices.md]: needed . These are => needed; they are:

File: docs/development_best_practices.md [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+No dynamic partitioning is needed, see next point about dynamic partitioning. 
+There is no distinct initial state for the partitions, i.e., all partitions start with the same initial state submitted during application launch. Typically input or output operators will not fall into this category, although there are exceptions and mainly it is operators in the middle of the DAG that are in this category.
+When used with non-input operators, only the data for the first declared input port is distributed among the different partitions. All other input ports are treated as broadcast and all partitions receive all the data for that port.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: This doesn't fit the pattern of being a "condition" that the operator satisfies (the two earlier points do). Suggest rewording.

File: docs/development_best_practices.md [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+* **StatelessThroughputBasedPartitioner** in Malhar provides a dynamic partitioner based on throughput thresholds. Similarly **StatelessLatencyBasedPartitioner** provides a latency based dynamic partitioner in RTS. If these partitioners can be used, then separate partitioning related code is not needed. The conditions under which these can be used are
```
Comment (amberarrow) [inline docs/development_best_practices.md]: are => are:

File: docs/development_best_practices.md [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+	* There is no distinct initial state for the partitions.
+	* There is no state change in the operator across windows i.e., operator is stateless.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Suggest rewording something like this: The operator maintains no state that needs to carry over from one window to the next.

File: docs/development_best_practices.md [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+
+### Custom partitioning
+
+In many cases, operators don’t satisfy the above conditions and a built-in partitioner cannot be used. Custom partitioning code needs to be written by the operator developer. Below are guidelines for it
```
Comment (amberarrow) [inline docs/development_best_practices.md]: for it => for it.

File: docs/development_best_practices.md [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+
+*StreamCodec* is also useful in another aspect of the application. It is used to serialize and deserialize the tuple to transfer it between operators. The default *StreamCodec* uses Kryo library for serialization. 
+
+The following guidelines are useful when considering a custom *StreamCodec*
+
+* A custom *StreamCodec* is needed if the tuples need to be distributed based on a criteria different from the hashcode of the tuple. If the correct working of an operator depends on the data from the upstream operator being distributed using a custom criteria such as being sticky on a “key” field within the tuple, then a custom *StreamCodec* should be provided by the operator developer. This codec can implement the custom criteria. The operator should also return this custom codec in the getStreamCodec method of the input port.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: highlight getStreamCodec in some way

File: docs/development_best_practices.md [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+
+## Threads
+
+The operator lifecycle methods such as **setup**, **beginWindow**, **endWindow**, **process** in *InputPorts* all are called from a single operator lifecycle thread unbeknownst to the user. So the user does not have to worry about dealing with the issues arising from multi-threaded code. Use of separate threads in an operator is discouraged because in most cases the motivation for this is parallelism, but parallelism can already be achieved by using multiple partitions and furthermore mistakes can be made easily when writing multi-threaded code. When dealing with high volume and velocity data, the corner cases with incorrectly written multi-threaded code are encountered more easily and exposed. However, there are times when separate threads are needed, for example, when interacting with external systems the delay in retrieving or sending data can be large at times, blocking the operator and other DAG processing such as committed windows. In these cases the following guidelines must be followed strictly
```
Comment (amberarrow) [inline docs/development_best_practices.md]: followed strictly => followed strictly.

File: docs/development_best_practices.md [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+
+* Threads should be started in **activate** and closed down in deactivate. Wait till they finish in **deactivate**.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Rewrite: Threads should be started in **activate**; symmetrically, they should be waited for and joined or terminated in **deactivate**.

**Comment by pramodin** [comment]
Incorporated most of the comments, just a small number of them I left it in the original state, as I felt they were appropriate. Let me know if you need clarification on those.

Will close this pull request shortly as I have submitted the latest with these changes in core, the pull request mentioned below. If you have further comments, please add them there.

https://github.com/apache/apex-core/pull/382
